src/main/java/com/google/api/codegen/discovery/MainDiscoveryProviderFactory.java

```diff
@@ -49,48 +49,61 @@
   public static final String PYTHON = "python";
   public static final String RUBY = "ruby";
 
-  private static final String DEFAULT_SNIPPET_FILE = "discovery_fragment.snip";
+  private static final String SNIPPET_FILE_ADC = "discovery_fragment.snip";
+  private static final String SNIPPET_FILE_3LO = "discovery_fragment_3lo.snip";
+  private static final String SNIPPET_FILE_API_KEY = "discovery_fragment_api_key.snip";
```

Comment:
But can we factor our the common parts so that we don't have code duplication? I could imagine two ways of doing it:

1. have a main template file for each auth mechanism. Each of these files consists mostly of the auth boiler plate, and at the bottom it imports a common template for the actual method invocation. This is the way this PR seems currently heading.

1. have a common template that imports the correct boilerplate depending on the auth mechanism used.

Both of these should prevent code duplication. Since the boilerplate is so long, it might be clearer to go for the first option, but I don't have strong feelings one way or the other at the moment. What do you guys think?